integration/run_test.go

```diff
@@ -185,3 +191,216 @@ func testWebServiceContainer(shell *shell.Shell, containerName string) {
+				}
+				var logLine ErrorLogLine
+				if err := json.Unmarshal([]byte(rawL), &logLine); err == nil {
+					if logLine.Msg == "statFile error" && strings.Contains(logLine.Error, "giving up after") {
+						gaveUpChannel <- true
+						return
+					}
+				}
+			}
+
+			m.Add("retry", monitorGaveUp)
+			defer m.Remove("retry")
+
+			// Do something in the container to access un-fetched spans, requiring network access
+			commandNetworkStdout, _, err := sh.R("ctr", "task", "exec", "--exec-id", "test-task-2", "test-container", "cat", "/etc/hosts")
+			if err != nil {
+				t.Fatalf("attempt to run task requiring network access failed: %s", err)
+			}
+
+			// Wait with network disabled
+			time.Sleep(time.Duration(tt.config.NetworkDisableMsec) * time.Millisecond)
+
+			// Short wait to allow commands to complete
+			time.Sleep(time.Duration(1000) * time.Millisecond)
+
+			// Confirm first command succeeded while network was down
+			buf := make([]byte, 100)
+			_, err = commandSucceedStdout.Read(buf)
+			if err != nil {
+				t.Fatalf("read from expected successful task output failed: %s", err)
+			}
+			if !strings.Contains(string(buf), "s ") {
+				t.Fatalf("expected successful task produced unexpected output: %s", string(buf))
+			}
+
+			// async read from command_network_stdout
+			commandNetworkStdoutChannel := make(chan []byte)
+			commandNetworkErrChannel := make(chan error)
+			go func() {
+				defer close(commandNetworkStdoutChannel)
+				defer close(commandNetworkErrChannel)
+				for {
+					var b []byte
+					if _, err := commandNetworkStdout.Read(b); err != nil && err != io.EOF {
+						commandNetworkErrChannel <- fmt.Errorf("read from network bound task output failed: %s", err)
+						return
+					}
+					commandNetworkStdoutChannel <- b
+					if err == io.EOF {
+						commandNetworkErrChannel <- fmt.Errorf("read from network bound task output encountered EOF")
+						return
+					}
+				}
+			}()
+
+			if tt.config.NetworkDisableMsec > 0 {
+
+				// Confirm second command has not succeeded while network was down
+				select {
+				case data := <-commandNetworkStdoutChannel:
+					t.Fatalf("network bound task produced unexpected output: %s", string(data))
+				case <-time.After(100 * time.Millisecond):
+				}
+
+				// Restore access to the registry and image
+				sh.X("iptables", "-D", "OUTPUT", "-d", registryHostIP, "-j", "DROP")
+
+				// Wait with network enabled, so a final retry has a chance to succeed
+				time.Sleep(2 * time.Millisecond * time.Duration(math.Min(
+					float64(tt.config.MaxWaitMsec),
+					math.Pow(2, float64(tt.config.MaxRetries))*float64(tt.config.MinWaitMsec),
+				)))
+			}
+
+			// Confirm whether second command has succeeded with network restored
+
+			select {
+			case gaveUp := <-gaveUpChannel:
+				if tt.config.ExpectedSuccess && gaveUp {
+					t.Fatal("retries gave up despite test expecting retry success")
+				}
+			case <-time.After(100 * time.Millisecond):
```

Comment:
Why do we need 100ms here? Can we just have one select over all the cases?

Reply:
The goroutine that would emit on `commandNetworkStdoutChannel` will sometimes not have caught up by this point, we have to give it time to read some output and process it.
That said, I can un-nest the selects and move the timeout down to the currently nested default.